### PR TITLE
Fix for empty results when no shards can be found to route to [R13, v3]

### LINKED ISSUE #10144: Bug Report: exists may returns random result on a sharded table
### Overview of the Issue

We found an issue with exists : I can't provide the exact query we use, but for the query like 
```
select exists(select * from foo where some_field=some_value)
```
here foo is in a sharded keyspace, it seems if the inner query has result in one shard but have no result in another shard, then the outer query will return a random 0 or 1.  Not sure if it is related, but for our particular query, the table is sharded on the field 'some_field', so if Vitess is smart enough, it doesn't need to send the query to all the shard. 
We tried using gen4 planner, got the same result.

### Reproduction Steps

See description.

### Binary Version

```sh
v13.0.0
```

### Operating System and Environment details

```sh
# cat /etc/os-release
PRETTY_NAME="Debian GNU/Linux 10 (buster)"
NAME="Debian GNU/Linux"
VERSION_ID="10"
VERSION="10 (buster)"
VERSION_CODENAME=buster
ID=debian
HOME_URL="https://www.debian.org/"
SUPPORT_URL="https://www.debian.org/support"
BUG_REPORT_URL="https://bugs.debian.org/"
```

### Log Fragments

_No response_

### PR DESCRIPTION
**Pull request by systay**
## Description
Queries that follow the following pattern - 
```mysql
select exists(select * from foo where some_field=some_value)
```

The table `foo` would has a lookup vindex on `some_field` and there are no matching rows for `some_value`. What Vitess would do when no shards could be found to route to is to return an empty result. This is correct for most cases, but not always. The query above would return an empty result instead of a `0`.

This fix is for the V3 planner.

## Related Issue(s)
Fixes #10144

## Checklist
-   [x] "Backport me!" label has been added if this change should be backported
-   [x] Tests were added or are not required
-   [x] Documentation was added or is not required

### CODE REVIEW

File: go/vt/vtgate/planbuilder/plan_test.go [hunk 1]
```diff
@@ -411,6 +422,21 @@ func TestWithDefaultKeyspaceFromFileSharded(t *testing.T) {
...
+		keyspace: &vindexes.Keyspace{
+			Name:    "user",
+			Sharded: true,
+		},
+		tabletType: topodatapb.TabletType_PRIMARY,
+	}
+
+	testOutputTempDir := makeTestOutput(t)
+	testFile(t, "select_cases_with_user_as_default.txt", testOutputTempDir, vschema)
+}
```
Comment (frouioui) [inline go/vt/vtgate/planbuilder/plan_test.go:438]: I might be missing, but I don't understand the clear difference between `TestOneWithUserAsDefault` and `TestWithUserDefaultKeyspaceFromFileSharded`. Is `select_cases_with_user_as_default.txt` a file where we will keep adding plan tests, or a file that serves for one-case test (like `onecase.txt`)?

File: go/vt/vtgate/planbuilder/route.go [hunk 2]
```diff
@@ -466,7 +472,7 @@ func (rb *route) SubqueryCanMerge(pb *primitiveBuilder, inner *route) bool {
 
 	// if either side is a reference table, we can just merge it and use the opcode of the other side
 	if rb.eroute.Opcode == engine.Reference || inner.eroute.Opcode == engine.Reference {
-		return true
+		return rb.isSingleShard() && inner.isSingleShard()
```
Comment (frouioui) [inline go/vt/vtgate/planbuilder/route.go:475]: We should update the comment

**Review by frouioui — APPROVED** [review]
There are failing tests, the expected output should be changed. Otherwise, looks good to me, I left a few nits and questions.

File: go/vt/vtgate/planbuilder/plan_test.go (hunk 1 above)
Comment (systay) [reply go/vt/vtgate/planbuilder/plan_test.go:438]: yes. `TestOne...` is for one-offs, and `TestWithUserDefaultKeyspaceFromFileSharded` is forever

**Pull request merged**